### PR TITLE
README: add architecture and directory overview

### PR DESCRIPTION
**Pull request by blackboxsw**
Add more context around to ua-client architecture and code.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -113,7 +185,7 @@ tox -e behave -D reuse_container=container_name
 
 ## Building
 
-The packaging for the UA client package (ubuntu-advantage-tools) is
```
Comment (patriciadomin) [inline README.md:116]: Thanks for updating this. Why are you just mentioning the -pro package?

File: README.md [hunk 2]
```diff
@@ -15,8 +15,80 @@ following entitlements are supported:
...
+
+## Architecture
+Ubuntu Advantage client, hereafter "UA client", is python3-based command line utility. It provides a CLI to attach, detach, enable,
+disable and check status of support related services.
+
+The package `ubuntu-advantage-tools` also provides a C++ APT hook which helps advertise ESM service and available packages in MOTD and during various apt commands.
+
+The `ubuntu-advantage-pro` package delivers auto-attach auto-enable functionality via init scripts and systemd services for various cloud platforms.
+
+By default, Ubuntu machines are deployed in an unattached state. A machine can get manually or automatically attached to a specific contract by interacting with the Contract Server REST API to enable a machine. Any change in state of services or machinbe attach results in additional interactions with the Contract Server API to validate such operations. The contract server API is described by the [ua-contracts openapi spec](https://github.com/CanonicalLtd/ua-contracts/blob/develop/docs/contracts.yaml).
```
Comment (lucasmoura) [inline README.md:43]: Maybe we can change the sentence from: 

`A machine can get manually or automatically attached to a specific contract by interacting with the Contract Server REST API to enable a machine`

To:

`A machine can get manually or automatically attached to a specific contract by interacting with the Contract Server REST API`

Also, there is a typo on `machinbe`

File: README.md [hunk 3]
```diff
@@ -15,8 +15,80 @@ following entitlements are supported:
...
+
+### Attaching a machine
+Each Ubuntu SSO account holder has access to one one or more contracts. To attach a machine to an Ubuntu Advantage contract:
```
Comment (lucasmoura) [inline README.md:46]: s/one one/one

File: README.md [hunk 4]
```diff
@@ -15,8 +15,80 @@ following entitlements are supported:
...
+
+* An Ubuntu SSO account holder must obtain a contract token from https://ubuntu.com/advantage.
+* Run `ua attach <contractToken>` on the machine or provide comparable
```
Comment (lucasmoura) [inline README.md:49]: Sorry, I don't know if I am missing something, but I did not understand `or provide comparable`  means in that context. Is it related to the machine or contract token ?

File: README.md [hunk 5]
```diff
@@ -15,8 +15,80 @@ following entitlements are supported:
...
+* UA Client reads config from /etc/ubuntu-advantage/uaclient.conf nad  a request containing the `<contractToken>` to the Contract Server API @ https://<contract_url>/api/v1/context/machines/token and obtains an unique machine-token JSON response containing credentials, affordances, directives and obligations to allow enabling and disabling any Ubuntu Advantage service
```
Comment (lucasmoura) [inline README.md:50]: s/nad/and

File: README.md [hunk 6]
```diff
@@ -15,8 +15,80 @@ following entitlements are supported:
...
+* UA Client POSTs a request containing the `<contractToken>` to the Contract Server API @ https://contracts.canonical.com/api/v1/context/machines/tokento to obtain an unique machine-token JSON response which contains credentials, affordances, directives and obligations to allow enabling and disabling any Ubuntu Advantage service
```
Comment (lucasmoura) [inline README.md:51]: I think you can remove the ending `to` from the url `https://contracts.canonical.com/api/v1/context/machines/tokento` 

Also. maybe I am missing something, but this paragraph seems really similar to the last one. I believe they are different ways to obtain the same JSON response and if that is the case, maybe we could address it in the docs. But it possible that | did not understand it correctly

File: README.md [hunk 7]
```diff
@@ -15,8 +15,80 @@ following entitlements are supported:
...
+POST `<contractToken>` to contracts server API @ https://contracts.canonical.com/api/v1/context/machines/token
+* UA client writes the API response to the root-readonly /var/lib/ubuntu-advantage/machine-token.json
+
+### Enabling a service
+Each service controlled by UA client will have a python module in uaclient/entitlements/\*.py which handles setup and teardown of services when enabled or disabled.
+
+If a contract entitles a machine to a service, `root` user can enable the service with `ua enable <service>`.  If a service can be disabled `ua disabled <service>` will be permitted.
+
+The goal of the UA client is to remain simple and flexible and let the
+contracts backend drive dynamic changes in contract offerings and contraints.
```
Comment (lucasmoura) [inline README.md:62]: s/contraints/constraints

File: README.md [hunk 8]
```diff
@@ -15,8 +15,80 @@ following entitlements are supported:
...
+In pursuit of that goal, the UA client obtains most of it's service contraints
```
Comment (lucasmoura) [inline README.md:63]: s/contraints/constraints

File: README.md [hunk 9]
```diff
@@ -15,8 +15,80 @@ following entitlements are supported:
...
+from a machine token that it ob
```
Comment (lucasmoura) [inline README.md:64]: I think this paragraph is incomplete

File: README.md [hunk 10]
```diff
@@ -15,8 +15,80 @@ following entitlements are supported:
...
+
+The UA Client is simple in that it relies on the machine token on the attached
+machine to describe whether a service is applicable for an envronment and what
+configuration is required to properly enable that service.
+
+Any interactions with the Contract server API are defined as UAContractClient class methods in uaclient/contract.py.
```
Comment (lucasmoura) [inline README.md:70]: Maybe we could put a direct link here to the python file

File: README.md [hunk 11]
```diff
@@ -15,8 +15,80 @@ following entitlements are supported:
...
+the manpage.
+
+## Terminology
+ The following vocabulary is used to describe different aspects of the work Ubuntu Advantage Client performs:
+
+| Term | Meaning |
+| -------- | -------- |
+| UA Client | The python command line client represented in this ubuntu-advantage-client repository. It is installed on each Ubuntu machine and is the entry-point to enable any Ubuntu Advantage commercial service on an Ubuntu machine. |
+| Contract Server | The backend service exposing a REST API to which UA Client authenticates in order to obtain contract and commercial service information and manage which support servics are active on a machine.|
+| Entitlement/Service | An Ubuntu Advantage commercial support service such as FIPS, ESM, Livepatch, CIS-Audit to which a Contract may be entitled |
```
Comment (TheRealFalcon) [inline README.md:29]: nit: Contract should probably be lowercase

File: README.md (hunk 4 above)
Comment (TheRealFalcon) [inline README.md:49]: Should probably specify run using sudo or as root

File: README.md [hunk 12]
```diff
@@ -15,8 +15,80 @@ following entitlements are supported:
...
+| uaclient.cli | The entry-point for the command-line client
+| uaclient.clouds | Cloud-platform detection logic used in Ubuntu Pro to determine if a given should be auto-attached to a contract |
+| uaclient.contract | Module for interacting with the Contract Server API |
+| ./demo | Various stale developer scripts for setting up one-off demo environemnts. (Not needed often)
+| /etc/ubuntu-advantage/uaclient.conf | Configuration file for the UA client.|
+| apt-hook | the C++ apt-hook delivering MOTD and apt command notifications about UA support services |
+| apt-conf.d | apt config files delivered to /etc/apt/apt-conf.d to automatically allow unattended upgrades of ESM  security-related components |
+| /var/lib/ubuntu-advantage/private | `root` read-only directory containing Contract API responses, machine-tokens and service credentials |
+ | /var/log/ubuntu-advantage.log | `root` read-only log of ubuntu-advantage operations |
+
```
Comment (TheRealFalcon) [inline README.md:91]: For this section in general, just be consistent with the `./`
E.g., .`./tools`, `.features/`, and `uaclient/` are all different even though they all refer to directories.

File: README.md [hunk 13]
```diff
@@ -113,23 +185,45 @@ tox -e behave -D reuse_container=container_name
 
 ## Building
 
-The packaging for the UA client package (ubuntu-advantage-tools) is
-in-tree, so you can build the package the way you would normally build
-a Debian package:
+The packaging for the UA client packages ubuntu-advantage-tools and
+ubuntu-advantage-pro) is in-tree in the debian directory. So you can build the
```
Comment (TheRealFalcon) [inline README.md:189]: `)` is now unnecessary

**Comment by TheRealFalcon** [comment]
Oh, also, I know you didn't update this section, but two of the links at the very top 404:
https://ubuntu.com/cis
https://ubuntu.com/fips-updates

File: README.md (hunk 4 above)
Comment (blackboxsw) [reply README.md]: oops, I had meant to discuss Ubuntu PRO here and didn't clarify that. 
Fixed to 
```
* Run `ua attach <contractToken>` on the machine                                
  - Ubuntu Pro images for AWS and Azure perform an auto-attach without tokens
```

Comment (blackboxsw) [reply README.md]: added sudo

File: README.md (hunk 5 above)
Comment (blackboxsw) [reply README.md]: +1

File: README.md (hunk 6 above)
Comment (blackboxsw) [reply README.md]: @lucasmoura Sorry I had accidentally double added that content. I've restructure the text and alignment and cleaned up a lot of inconsistencies now.

File: README.md (hunk 7 above)
Comment (blackboxsw) [reply README.md]: fixed.

File: README.md (hunk 8 above)
Comment (blackboxsw) [reply README.md]: fixed

File: README.md (hunk 9 above)
Comment (blackboxsw) [reply README.md]: Fixed ... "from a machine token that it obtains from the Contract Server API. "

File: README.md (hunk 10 above)
Comment (blackboxsw) [reply README.md]: +1

File: README.md (hunk 12 above)
Comment (blackboxsw) [reply README.md:119]: thanks, did this and moved all the non-relative file paths mentioned to the bottom.

File: README.md (hunk 1 above)
Comment (blackboxsw) [reply README.md:116]: I also added some context above in the intro
```
Ubuntu Pro images on AWS and Azure Ubuntu Pro 
images will also contain `ubuntu-advantage-pro` which automates machine attach  
on custom AWS and Azure images. 
```

File: README.md (hunk 13 above)
Comment (blackboxsw) [reply README.md]: I rewrote my ramble here :/.
Thanks, it now says:
```
Creating ubuntu-advantage-tools and ubuntu-advantage-pro is created from the    
debian/control file in this repository. You can build the                       
package the way you would normally build a Debian package:       
```

**Comment by blackboxsw** [comment]
> Oh, also, I know you didn't update this section, but two of the links at the very top 404:
> https://ubuntu.com/cis
> https://ubuntu.com/fips-updates

Addressed, and set them to cis-audit and fips respectively (just like the `ua --help` command output)

File: README.md [hunk 14]
```diff
@@ -15,8 +15,108 @@ following entitlements are supported:
...
+Users can manually run the `ua` command to learn more or view the manpage.
+
+## Terminology
+ The following vocabulary is used to describe different aspects of the work
+Ubuntu Advantage Client performs:
+
+| Term | Meaning |
+| -------- | -------- |
+| UA Client | The python command line client represented in this ubuntu-advantage-client repository. It is installed on each Ubuntu machine and is the entry-point to enable any Ubuntu Advantage commercial service on an Ubuntu machine. |
+| Contract Server | The backend service exposing a REST API to which UA Client authenticates in order to obtain contract and commercial service information and manage which support servics are active on a machine.|
```
[the remark this answers is not in the document]
Comment (blackboxsw) [reply README.md:32]: done.

File: README.md [hunk 15]
```diff
@@ -15,8 +15,108 @@ following entitlements are supported:
...
+a machine to an Ubuntu Advantage contract:
+
+* An Ubuntu SSO account holder must obtain a contract token from
+https://ubuntu.com/advantage.
+* Run `sudo ua attach <contractToken>` on the machine
+  - Ubuntu Pro images for AWS and Azure perform an auto-attach without tokens
+* UA Client reads config from /etc/ubuntu-advantage/uaclient.conf to obtain
+  the contract_url (default: https://contracts.canonical.com)
+* UA Client POSTs to the Contract Server API @
+  <contract_url>/api/v1/context/machines/token providing the <contractToken>
```
[the remark this answers is not in the document]
Comment (blackboxsw) [reply README.md:70]: thanks, and done.

File: README.md [hunk 16]
```diff
@@ -15,8 +15,108 @@ following entitlements are supported:
...
+| File/Directory | Intent |
+| -------- | -------- |
+| ./tools | Helpful scripts used to publish, release or test various aspects of UA client |
+| ./features/ | Behave BDD integration tests for UA Client
+| ./uaclient/ | collection of python modules which will be packaged into ubuntu-advantage-tools package to deliver the UA Client CLI |
+| uaclient.entitlements | Service-specific \*Entitlement class definitions which perform enable, disable, status, and entitlement operations etc. All classes derive from base.py:UAEntitlement and many derive from repo.py:RepoEntitlement |
+| ./uaclient/cli.py | The entry-point for the command-line client
+| ./uaclient/clouds/ | Cloud-platform detection logic used in Ubuntu Pro to determine if a given should be auto-attached to a contract |
+| uaclient.contract | Module for interacting with the Contract Server API |
+| ./demo | Various stale developer scripts for setting up one-off demo environemnts. (Not needed often)
```
[the remark this answers is not in the document]
Comment (blackboxsw) [reply README.md:113]: +1

File: README.md [hunk 17]
```diff
@@ -7,16 +7,116 @@ view, enable, and disable offerings from Canonical on their system. The
...
+service with `ua enable <service>`.  If a service can be disabled
+`ua disabled <service>` will be permitted.
+
+The goal of the UA client is to remain simple and flexible and let the
+contracts backend drive dynamic changes in contract offerings and constraints.
+In pursuit of that goal, the UA client obtains most of it's service constraints
+from a machine token that it obtains from the Contract Server API.
+
+The UA Client is simple in that it relies on the machine token on the attached
+machine to describe whether a service is applicable for an envronment and what
```
Comment (patriciadomin) [inline README.md:94]: typo environment

Comment (blackboxsw) [reply README.md]: thanks and Fixed.

**Pull request merged**